Tuples: adding support for tuples 8 and up

jcouv opened the pull request:
`TupleTypeSymbol` now holds a chain of underlying types, instead of just one. 
For instance, it will keep a chain such as `ValueTuple<....., ValueTuple<string>>` and `ValueTuple<string>` to represent the types underlying `(7x int, string)`.

Similarly, `TupleFieldSymbol` now holds a chain of underlying field symbols, instead of just one. 
This allows an access the string above using `(tuple).Item8` or the element's name to be converted to `ValueTuple.Rest.Item1`.

Because the common case is short tuples, those chains use `OneOrMany`.

The binder (`Binder_Expression.cs` and `Binder_Symbols.cs`) now allows binding the declaration and creation of a tuple type with more than 7 elements.

Local lowering (`LocalRewriter_Field.cs` and `LocalRewriter_TupleCreationExpression.cs`) rewrites tuple field access and tuple creation to match the nesting of `ValueTuple` types.

@dotnet/roslyn-compiler for review.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -73,31 +82,154 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+                    for (var linkIndex = 0; linkIndex < numLinks; linkIndex++)
+                    {
+                        var linkArity = (linkIndex != maxNumLinks - 1) ? 8 : SizeOfLastValueTuples(numElements);
+                        var linkPosition = (linkIndex != numLinks - 1) ? 7 : elementIndex % 7;
```
There are many instances of `7` here and below. Consider creating a `const`.

jcouv replied:
I didn't address this yet. I think fewer constant `7` appear after I address the first round of PR feedback. Let me know if you still think it is worthwhile.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -73,31 +82,154 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+            var fieldsBuilder = ArrayBuilder<TupleFieldSymbol>.GetInstance(numElements);
+
+            var hasChunks = numElements > 7;
+            if (!hasChunks)
```
Consider inlining `hasChunks`.

jcouv replied:
Gone in new code.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -73,31 +82,154 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+                                               elementTypes[elementIndex],
+                                               OneOrMany.Create(underlyingField?.AsMember(underlyingTypeChain[0])));
+                    fieldsBuilder.Add(field);
+                }
```
Is the short tuple not handled by the long tuple code below?

jcouv replied:
It is, and I would like to merge the two cases, but I'm trying to avoid creating an array of one, since short tuples are the common case.

AlekseyTs replied:
I think maintenance of the field chain is complicating things, things should be simpler once we get rid of it.

jcouv replied:
Avoiding storing the chains indeed simplifies the code. There is no longer a special casing of one vs. many.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -73,31 +82,154 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+        internal ImmutableArray<MethodSymbol> GetTupleConstructorChain(TupleExpressionSyntax node, Binder binder, DiagnosticBag diagnostics, ref bool hasErrors)
+        {
+            var nestings = TupleTypeSymbol.NumberOfValueTuples(Arity);
+            int nextArity = TupleTypeSymbol.SizeOfLastValueTuples(Arity);
```
Are we usually calling `NumberOfValueTuples` and `SizeOfLastValueTuples` together? If so, consider returning from a single `NumberOfAndRemainder` method.

jcouv replied:
Good idea. I'll do that.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -43,24 +40,36 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol, ITupleTypeSymbol
+            for (int i = 0; i < newUnderlyingTypeChain.Count; i++)
+            {
+                Debug.Assert((object)newUnderlyingTypeChain[i].OriginalDefinition == (object)UnderlyingTupleTypeChain[i].OriginalDefinition);
+            }
```
Wrap loop in ```#if DEBUG``` or use:
```
Debug.Assert(newUnderlyingTypeChain.Select(t => t.OriginalDefinition).SequenceEqual(UnderlyingTupleTypeChain.Select(t => t.OriginalDefinition));
```

jcouv replied:
The loop is gone now that I don't store the chain.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -73,31 +82,154 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+            if (numChunks == 1)
+            {
+                var tupleType = binder.GetWellKnownType(GetTupleType(nextChunkSize), diagnostics, syntax);
+                var symbol = tupleType.Construct(ImmutableArray.Create(elementTypes, 0, nextChunkSize));
```
```tupleType.Construct(elementTypes)```

cston replied:
```
tupleType.Construct(elementTypes)
```

jcouv replied:
Will do. Thanks.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -107,10 +239,9 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
-                // PROTOTYPE
-                arity = 1;
+                throw ExceptionUtilities.Unreachable;
             }
```
Is the ```if (arity > 8) throw Unreachable;``` necessary? Indexing into `tupleTypes` will already throw if out of range. If it's necessary consider using ```if (arity > tupleTypes.Length)``` instead. Same comment for `GetTupleCtor` below.

AlekseyTs replied:
I would argue that Unreachable is preferred over IndexOutOfRange

jcouv replied:
I left `Unreachable`.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -427,20 +573,20 @@ private bool Equals(TupleTypeSymbol other, bool ignoreCustomModifiers, bool igno
         public override int GetHashCode()
         {
-            return _underlyingType.GetHashCode();
+            return _underlyingTypeChain.GetHashCode();
```
```OneOrMany<T>``` does not define `GetHashCode`.

jcouv replied:
Not applicable after removing the chain.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -20,7 +17,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
     internal sealed class TupleTypeSymbol : NamedTypeSymbol, ITupleTypeSymbol
     {
-        private readonly NamedTypeSymbol _underlyingType;
+        private readonly OneOrMany<NamedTypeSymbol> _underlyingTypeChain;
```
Is ```_underlyingTypeChain[0]``` the constructed type for ```ValueTuple<T1, ..., T7, ValueTuple<T8, ..., T15, ValueTuple<...>>>```? If so, doesn't that single constructed type hold the chain of underlying types?

AlekseyTs replied:
I agree, this is an overkill, we don't need this chain because the very first (top-level) underlying tuple type has all the information already. Having the chain doesn't simplify anything, it uses more memory and requires more maintenance like with generic substitution. With generic substitution, for example we are going to substitute the same types multiple time, the longer the chain, more duplication will be there. Given a chain of length N, the last type in the chain will be substituted N times, the one before that (N-1) times, etc. Whereas, all we need to substitute is the first (top-level) type, then all types in the chain will be substituted only once.

jcouv replied:
Makes sense. I will extract the nested types from the largest type.

AlekseyTs commented:
For (7x int, string) tuple, why not use a single tuple8 type and put string into the eighth field?

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -200,20 +335,27 @@ private static WellKnownMember GetTupleTypeMember(int arity, int position)
-            }
-            return memberNames[position];
+            var number = position + 1;
+            return "Item" + number;
```
By analogy, shouldn't we reserve all names that fit the pattern of "Item" + decimal integer without leading zeros?

jcouv replied:
I think so, but didn't want to touch the reserved name logic just yet (we know it has to be re-done). 
The current reserved name logic should be aware of methods from ValueTuple (`Item7`, `Rest`, `ToString`). I'm not sure how that will be done yet.

VSadov replied:
It seems Item9, Item10 and so on should not be reserved. They do not clash with anything. 
We should consider allowing Item9, just like we allow ITEM9, item9, Item_9, Item09  etc.. anyways.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -226,11 +368,15 @@ internal static bool IsMemberNameReserved(string name)
         }
         private static readonly string[] memberNames = { "Item1", "Item2", "Item3", "Item4", "Item5", "Item6", "Item7" };
```
Is this a good time to reserve "Rest"?

jcouv replied:
Sure. I'll add.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -659,7 +660,7 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
                     countOfExplicitNames++;
 
-                    if (TupleTypeSymbol.IsMemberNameReserved(name) && name != TupleTypeSymbol.UnderlyingMemberName(i))
+                    if (TupleTypeSymbol.IsMemberNameReserved(name) && name != TupleTypeSymbol.TupleMemberName(i))
```
It feels like we should reserve "Rest" as a name at any position. In addition we can simplify this check and avoid unnecessary string concatenations and comparisons if we modify IsMemberNameReserved to return the index, at which reserved name can be used, -1 if not reserved. Then ```name != TupleTypeSymbol.TupleMemberName(i)``` condition can be simplified to ```<reserved index> != i```.

VSadov replied:
The real reason for "reserving" the name is that it should not clash with what already exists per our own definition of the Well Known Members of this type.

The issue with "Rest" is that it would not clash in any position because of the forced nesting of item N8. 

Perhaps we should actually allow it at any position since there is no possible clash?

VSadov replied:
Ok, discussed this further with Aleksey. Because of unavoidable shadowing due to wrapping (which would be very confusing), it is better to not allow "Rest" in any position.

jcouv replied:
We have to allow `Rest` at position 8 (and make it function properly, meaning it would return the `Rest` field without unwrapping if it is a `ValueTuple`) for compatibility with C# 6 users of `ValueTuple`.

AlekseyTs replied:
The proposal is to disallow assigning name "Rest" to any item in the tuple, exactly for the purpose of disallowing shadowing of the "extension" field. Access tuple.Rest is still going to be allowed. What compatibility concerns do you have?

jcouv replied:
As long as access through `Rest` is allowed we're good. I will disallow naming a field as `Rest` as suggested.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -721,15 +716,7 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
-            {
-                hasErrors = true;
-            }
+            var ctor = type.GetTupleConstructorChain(node, this, diagnostics, ref hasErrors);
```
Please spell-out the type. In general, we disallow use of ```var``` unless the type is mentioned in the initializer. For example, "var builder = **ArrayBuilder\<int\>**.GetInstance();" is Ok, but this statement is not Ok. Usage of var in this manner negatively affects readability of the code.

AlekseyTs replied:
In addition, it looks like the name of the local no longer reflects what it is used for.

jcouv replied:
Thanks. I will update `var` in multiple places to follow this convention. Makes sense.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +10,43 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+            {
+                // replace receiver
+                return node.Update(rewrittenReceiver, node.FieldSymbol, node.ConstantValue, node.ResultKind, node.Type);
+            }
```
It doesn't look like there is a good reason to allocate a node that will be right away dropped on the floor. I think we shouldn't even call this method for non-tuple case.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +10,43 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+            if (tupleField.UnderlyingFieldSymbolChain.Count == 1)
+            {
+                // short tuple: replace receiver and field symbol
+                return node.Update(rewrittenReceiver, tupleField.UnderlyingFieldSymbolChain[0], node.ConstantValue, node.ResultKind, node.Type);
```
This should call MakeFieldAccess and return the final node.

AlekseyTs replied:
If we are confident that we can build the final node without calling MakeFieldAccess, that is fine, but we should produce a node that doesn't have to be lowered more during this lowering pass.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +10,43 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+            for (var i = 0; i < chainLength; i++)
+            {
+                var nestedFieldType = (i < chainLength - 1) ? tuple.UnderlyingTupleTypeChain[i + 1] : node.Type;
+                nestedField = new BoundFieldAccess(node.Syntax, nestedField, tupleField.UnderlyingFieldSymbolChain[i], node.ConstantValue, node.ResultKind, nestedFieldType);
```
Why is it appropriate to reuse node.ConstantValue, node.ResultKind for all new nodes, but the last one?

jcouv replied:
I didn't understand the question.

AlekseyTs replied:
Why are you using node.ConstantValue, node.ResultKind for intermediate fields? Why do you think it is the right thing to do?

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +10,43 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+            {
+                var nestedFieldType = (i < chainLength - 1) ? tuple.UnderlyingTupleTypeChain[i + 1] : node.Type;
+                nestedField = new BoundFieldAccess(node.Syntax, nestedField, tupleField.UnderlyingFieldSymbolChain[i], node.ConstantValue, node.ResultKind, nestedFieldType);
+            }
```
I think, if the code was written without conditional logic inside the loop, it would be more efficient and easier to understand:
```
            var tuple = (TupleTypeSymbol)tupleField.ContainingType;
            int countOfIntermediateFields = tupleField.UnderlyingFieldSymbolChain.Count - 1;
            for (var i = 0; i < countOfIntermediateFields; i++)
            {
                rewrittenReceiver = new BoundFieldAccess(node.Syntax, 
                                                         rewrittenReceiver,
                                                         tupleField.UnderlyingFieldSymbolChain[i],
                                                         ..., ..., 
                                                         tuple.UnderlyingTupleTypeChain[i + 1]);
            }

            return new BoundFieldAccess(node.Syntax, 
                                        rewrittenReceiver,
                                        tupleField.UnderlyingFieldSymbolChain[countOfIntermediateFields],
                                        node.ConstantValue, node.ResultKind, node.Type);
```

Also note that as an added benefit there will be no need to to special case ```tupleField.UnderlyingFieldSymbolChain.Count == 1``` scenario above.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +10,43 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+        /// For instance, tuple.Item8
+        /// produces fieldAccess(field=Item1, receiver=fieldAccess(field=Rest, receiver=ValueTuple for tuple))
+        /// </summary>
+        private BoundFieldAccess ExpandTupleAccessor(BoundFieldAccess node, BoundExpression rewrittenReceiver)
```
Consider following the established naming pattern and calling this method "RewriteTupleFieldAccess". Its result shouldn't require more post-processing during this pass, like calling MakeFieldAccess in the caller.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -14,7 +10,40 @@ internal partial class LocalRewriter
         {
             var rewrittenArguments = VisitList(node.Arguments);
-            return new BoundObjectCreationExpression(node.Syntax, node.ConstructorOpt, rewrittenArguments);
+            BoundObjectCreationExpression nextCreation = ExpandTupleConstructor(node, rewrittenArguments);
```
The local is unnecessary and has confusing name.

jcouv replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -14,7 +10,40 @@ internal partial class LocalRewriter
+        /// For instance, for a long tuple we'll generate:
+        /// creationExpression(ctor=largestCtor, args=firstArgs+(nested creationExpression for remainder, with smaller ctor and next few args))
+        /// </summary>
+        private static BoundObjectCreationExpression ExpandTupleConstructor(BoundTupleCreationExpression node, ImmutableArray<BoundExpression> rewrittenArguments)
```
RewriteTupleCreationExpression?

jcouv replied:
Yes.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -14,7 +10,40 @@ internal partial class LocalRewriter
+        /// </summary>
+        private static BoundObjectCreationExpression ExpandTupleConstructor(BoundTupleCreationExpression node, ImmutableArray<BoundExpression> rewrittenArguments)
+        {
+            var numChunks = TupleTypeSymbol.NumberOfValueTuples(rewrittenArguments.Length);
```
```int``` has same length as ```var```

jcouv replied:
I'll fix that throughout my change.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -14,7 +10,40 @@ internal partial class LocalRewriter
+                if (chunkIndex == numChunks - 1)
+                {
+                    var firstChunkSize = TupleTypeSymbol.SizeOfLastValueTuples(rewrittenArguments.Length);
+                    chainedArguments = ImmutableArray.Create(rewrittenArguments, chunkIndex * 7, firstChunkSize);
```
Consider asserting that ```chunkIndex * 7 + firstChunkSize == rewrittenArguments.Length```

AlekseyTs replied:
Also, does ImmutableArray.Create reuse original rewrittenArguments array when we have only one chunk?

jcouv replied:
I forgot to address the suggest to assert, but I think it is un-necessary. The two variables are computed from the original length and so will always add up properly.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -14,7 +10,40 @@ internal partial class LocalRewriter
+        {
+            var numChunks = TupleTypeSymbol.NumberOfValueTuples(rewrittenArguments.Length);
+
+            var nextCreation = default(BoundObjectCreationExpression);
```
I am not sure why the local is called "next" when it feels like it is "current"

jcouv replied:
Renamed to `currentXYZ` in various loops for consistency.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -14,7 +10,40 @@ internal partial class LocalRewriter
+            {
+                ImmutableArray<BoundExpression> chainedArguments;
+
+                if (chunkIndex == numChunks - 1)
```
Consider eliminating conditional logic from the loop. The part for ```chunkIndex == numChunks - 1``` should always be performed. So, let's move this out of the loop (before it, along with BoundObjectCreationExpression creation) and adjust the loop initialization accordingly. As an added benefit, we will be able to assign real value to nextCreation right at the declaration.

jcouv replied:
I tried simplifying the loops as much as possible in the second commit. I improved one or two, but two (in `LocalRewriter`) still have this type of conditional. Removing them would duplicate 4 or 5 lines of code outside the loop.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -43,24 +40,36 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol, ITupleTypeSymbol
-                                                              return new TupleFieldSymbol(field.Name, this, newUnderlyingField.Type, newUnderlyingField);
-                                                          });
+            {
+                var newUnderlyingFieldChain = field.UnderlyingFieldSymbolChain.SelectAsOneOrMany((link, arg) => link.OriginalDefinition.AsMember(arg), newUnderlyingTypeChain);
```
'link' and 'arg' are not meaningful names.

jcouv replied:
Gone

AlekseyTs commented on src/Compilers/Core/Portable/InternalUtilities/OneOrMany.cs:
```diff
@@ -35,6 +36,43 @@ public OneOrMany(ImmutableArray<T> many)
+        /// <summary>
+        /// Maps each item in this OneOrMany using the map function and the corresponding argument from args.
+        /// </summary>
+        internal OneOrMany<TResult> SelectAsOneOrMany<TArgs, TResult>(Func<T, TArgs, TResult> map, OneOrMany<TArgs> args)
```
TArg vs. TArgs

AlekseyTs replied:
Also the name of the method is confusing. This is not a Select, this is more like a Zip (see Enumerable.Zip<TFirst, TSecond, TResult>).

jcouv replied:
The name was by imitation with `ImmutableArray.SelectAsArray` (the one that takes args). But I agree. I'll rename to `ZipAsOneOrMany`.

jcouv replied:
Since I won't be using `OneOrMany` to store chains, I'll undo those changes.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -73,31 +82,154 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+        /// Returns the size of the last ValueTuple.
+        /// For example, for 9 elements, the last chunk's size is 2.
+        /// </summary>
+        static internal int SizeOfLastValueTuples(int numElements)
```
SizeOfLastValueTuple vs. SizeOfLastValueTuple**s**

jcouv replied:
Combined both methods into `NumberOfValueTuples` as Chuck suggested.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -73,31 +82,154 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+        /// For short tuples, a single constructor is returned.
+        /// For long tuples, a chain of constructors for nested types (largest first).
+        /// </summary>
+        internal ImmutableArray<MethodSymbol> GetTupleConstructorChain(TupleExpressionSyntax node, Binder binder, DiagnosticBag diagnostics, ref bool hasErrors)
```
It is an overkill to build and pass around the chain of constructors, though, we do have to verify that all of them are present and good.

AlekseyTs replied:
It looks like constructor is the good time to verify presence of the constructors.

jcouv replied:
I moved the constructor logic into `RewriteTupleCreationExpression` (part of `LocalRewriter`). If a constructor is missing it is able to report diagnostics properly. Let me know if that's ok.

AlekseyTs commented on src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml:
```diff
@@ -1144,8 +1144,8 @@
-    <!-- The constructor of the underlying tuple type -->
-    <Field Name="ConstructorOpt" Type="MethodSymbol" Null="allow"/>
+    <!-- The constructor(s) of the underlying tuple type. One for short tuples. Multiple for long tuples. (largest first) -->
+    <Field Name="ConstructorChainOpt" Type="ImmutableArray&lt;MethodSymbol&gt;" Null="allow"/>
```
The only time we need this information is when we lower this node and all constructors can be easily recomputed based on the underlying tuple type. I think we should get rid of this member.

VSadov replied:
We had issues when types were there, but constructors missing. Ideally one should get a missing member error during binding, so we need to check if ctor is present.
Now it becomes a burden to keep all the constructors around. So, yes, perhaps we should check for the presence of the ctors, but not keep them, and mrk node as hasErrors if any one is missing so we would not try to lower. . When needed ast lowering time, we can find the ctors again.

jcouv replied:
Done.

jcouv commented:
@AlekseyTs, regarding the usage of the 8th field:
We discussed this at length with Vlad, Neal and Don, and decided to adopt the same convention as `System.Tuple`. This makes the API more consistent and also helps the interop with F#. It will also help for interop between ref and value tuples if we decide to do it.

If you look at the `System.Tuple` code, the 8th element is called `TRest` and is always expected to be a `ValueTuple` (although there is no way to constrain that until we have a public ITuple interface).

For reference, [F# spec (section 6.3.2)](http://fsharp.org/specs/language-spec/4.0/FSharpSpec-4.0-latest.pdf#page=69).

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -73,31 +82,154 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+            var numElements = elementTypes.Length;
+            var numChunks = NumberOfValueTuples(numElements);
+            var nextChunkSize = SizeOfLastValueTuples(numElements);
+            if (numChunks == 1)
```
It feels like once we get rid of requirement to build the chain as a separate data structure, the code in this function could be significantly simplified and can use one loop.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -122,19 +253,23 @@ private static WellKnownType GetTupleType(int arity)
         /// </summary>
         internal static WellKnownMember GetTupleCtor(int arity)
         {
             if (arity > 7)
```
I would expect this condition to be adjusted and the body of the 'if' to throw.

jcouv replied:
I'm thinking to put a Debug.Assert instead.

AlekseyTs replied:
It doesn't feel appropriate to have the fallback since we now have full support for "long" tuples. I don't think an assert   would address this concern.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -701,19 +847,26 @@ internal sealed class TupleFieldSymbol : FieldSymbol
         /// but caller is supposed to report some kind of error at declaration.
         /// </summary>
-        public TupleFieldSymbol(string name, TupleTypeSymbol containingTuple, TypeSymbol type, FieldSymbol underlyingFieldOpt)
+        public TupleFieldSymbol(string name, TupleTypeSymbol containingTuple, TypeSymbol type, OneOrMany<FieldSymbol> underlyingFieldChainOpt)
```
It is an overkill to keep and pass around the chain of fields. All we need to know in about the field is its final underlying field, the top level underlying tuple type for the owning TupleTypeSymbol and the field ordinal within the TupleTypeSymbol. With that information we can easy rebuild the chain during lowering.

AlekseyTs replied:
Do we have a test that covers generic substitution of compound tuples?

AlekseyTs replied:
I am done with review pass.

AlekseyTs replied:
Actually, we don't even need to know the underlying field symbol until lowering, we just keep doing unnecessary work to adjust it during tuple generic substitution. Let's get rid of it.

VSadov replied:
We should not optimize for gigantic tuples (especially if at the cost to small ones).
Lets try just keep the top level type and do everything based off the field's ordinal position.

VSadov commented:
done with reviewing. I think removing the chain and keeping just top level underlying type will result in many changes and hopefully will make it all simpler.

jcouv commented:
@AlekseyTs @cston @VSadov I pushed a new commit addressing most feedback.

The tuple and tuple field symbols now hold much fewer references and chains. Same for the tuple creation bound node (removed the constructor chain). 
But I did leave one in tuple field symbol, to facilitate multiple properties.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -692,29 +779,55 @@ internal override ImmutableArray<NamedTypeSymbol> GetInterfacesToEmit()
     internal sealed class TupleFieldSymbol : FieldSymbol
```
Consider making this class a `private` member of `TupleTypeSymbol`, or the class is used elsewhere, move the class to a separate file.

jcouv replied:
Moved to a separate file.

cston commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -806,171 +959,194 @@ class C
+                // (6,181): error CS8202: Tuple membername 'Rest' is disallowed at any position.
+                //         (int Item1, string Item3, int Item2, int Item4, int Item5, int Item6, int Item7, int Rest) x = (Item2: 1, Item4: "hello", Item3: 3, Item4: 4, Item5: 5, Item6: 6, Item7: 7, Rest: 8);
+                Diagnostic(ErrorCode.ERR_TupleReservedMemberNameAnyPosition, "Rest").WithArguments("Rest").WithLocation(6, 181)
+               );
```
Are we testing `Item0`, `Item8`, `Item10`, `Item01`?

jcouv replied:
`Item01` indeed has a problem. Adding test.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -107,10 +158,9 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
         private static WellKnownType GetTupleType(int arity)
         {
-            if (arity > 7)
+            if (arity > 8)
```
There are multiple uses of `7` and `8`. Please replace with a `const`.

jcouv replied:
Ok

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -200,32 +254,65 @@ private static WellKnownMember GetTupleTypeMember(int arity, int position)
+        /// Checks whether the field name is reserved and tells us which position it's reserved for.
+        ///
+        /// For example:
+        /// Returns 3 for "Item3 in positions other than 3.
```
Missing a closing ```"```?

AlekseyTs replied:
Also "in positions other than 3" doesn't sound quite right

jcouv replied:
Yes. Fixed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -200,32 +254,65 @@ private static WellKnownMember GetTupleTypeMember(int arity, int position)
+        /// Returns "Item12" for position=11
         /// </summary>
-        internal static string UnderlyingMemberName(int position)
+        internal static string TupleMemberName(int position)
```
Consider if it would make sense to make TupleMemberName and IsMemberNameReserved consistent in terms of position numbers. IsMemberNameReserved uses one-based positions, TupleMemberName uses zero-based positions.

jcouv replied:
Makes sense. Fixed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -704,13 +711,7 @@ private BoundExpression BindTupleExpression(TupleExpressionSyntax node, Diagnost
-                diagnostics.Add(ErrorCode.ERR_PrototypeNotYetImplemented, node.Location);
-                return BadExpression(node);
-            }
+            ImmutableArray<TypeSymbol> elements = elementTypes.ToImmutableAndFree();
```
No change needed. But we do sort of get lazy and use ```var``` with widely used APIs that "everybody" knows what they return. I guess ToImmutableAndFree is one of them.

AlekseyTs replied:
To clarify "No change needed" means what you have now is good.

jcouv replied:
This discussion of `var` made me think we could have a code fixer for that (click on `var` and "Replace with actual type" ;-)

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -396,9 +396,14 @@ internal Symbol BindNamespaceOrTypeOrAliasSymbol(ExpressionSyntax syntax, Diagno
+            var elementNames = ArrayBuilder<string>.GetInstance(count);
+
+            // set of names already used
+            HashSet<string> uniqueFieldNames = new HashSet<string>();
```
Would it make sense to use pooled HashSet?

jcouv replied:
Sure. Done.

AlekseyTs replied:
Both places?

jcouv replied:
Yup.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -406,19 +411,64 @@ private TypeSymbol BindTupleType(TupleTypeSyntax syntax, DiagnosticBag diagnosti
+                    else if (!uniqueFieldNames.Add(name))
+                    {
+                        Error(diagnostics, ErrorCode.ERR_TupleDuplicateMemberName, argumentSyntax.Name.Identifier);
+                    }
```
This looks like a clone of code in BindTupleExpression. Consider extracting into a method and reusing in both places.

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -406,19 +411,64 @@ private TypeSymbol BindTupleType(TupleTypeSyntax syntax, DiagnosticBag diagnosti
+                }
+                else
+                {
+                    if (name != null)
```
It looks like we have a test gap, this code is not reachable because we allocate elementNames builder at the point of  variable declaration. Please close the gap and adjust the code appropriately.

jcouv replied:
There is a test specifically for this late initialization (see `TupleTypeWithLateDiscoveredName`). But you are correct that it was no longer hitting this case because we always initialized `elementNames` early.
Fixed.

I'm wondering if there is a better way to do this lazy initialization... I looked at `Lazy<T>`, but it doesn't seem very convenient for this context.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -406,19 +411,64 @@ private TypeSymbol BindTupleType(TupleTypeSyntax syntax, DiagnosticBag diagnosti
+                        }
+                        elementNames.Add(name);
+                    }
+                }
```
This chunk also looks like a clone. Also consider extracting and reusing.

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -406,19 +411,64 @@ private TypeSymbol BindTupleType(TupleTypeSyntax syntax, DiagnosticBag diagnosti
+
+                // add the name to the list
+                // names would typically all be there or none at all
+                // but in case we need to handle this in error cases
```
It looks like we are missing the code that reports ERR_TupleExplicitNamesOnAllMembersOrNone error, please add test.

jcouv replied:
Good catch. Test is there (`TupleWithOnlySomeNames`) but should expect this error on tuple declarations too.

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +10,65 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+            if (tupleField == null)
+            {
+                return MakeFieldAccess(node.Syntax, rewrittenReceiver, node.FieldSymbol, node.ConstantValue, node.ResultKind, node.Type, node);
+            }
```
Consider moving this case to the caller so that `RewriteTupleFieldAccess` is only called for tuple fields.

AlekseyTs replied:
+1

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +10,65 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+                FieldSymbol nestedFieldSymbol = linkField?.AsMember(currentLinkType);
+
+                TypeSymbol nestedFieldType = currentLinkType.TypeArguments[indexInLink - 1];
+                currentField = new BoundFieldAccess(node.Syntax, currentField, nestedFieldSymbol, node.ConstantValue, node.ResultKind, nestedFieldType);
```
`BoundFieldAccess` does not support null `nestedFieldSymbol` (in the error case). Please add a test.

jcouv replied:
We can't actually get `null` at that point because all the field accessors have been checked when constructing the symbol (see `GetTupleUnderlyingTypeAndFields`).
I'll change the code to check for `null` and return `node`, which is always safe.

AlekseyTs replied:
I would say that it would be unexpected to get here with null field then. An assert with a comment about checked earlier should be sufficient.

AlekseyTs replied:
It is not appropriate to use node.ConstantValue, node.ResultKind on iterations that don't handle the very last access.

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +10,65 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
 
-            return MakeFieldAccess(node.Syntax, rewrittenReceiver, field, node.ConstantValue, node.ResultKind, node.Type, node);
+            var expandedNode = (BoundFieldAccess)currentField;
+            return MakeFieldAccess(expandedNode.Syntax, expandedNode.ReceiverOpt, expandedNode.FieldSymbol, expandedNode.ConstantValue, expandedNode.ResultKind, expandedNode.Type, expandedNode);
```
Do we need to call `MakeFieldAccess` on the result? If so, should the `oldNodeOpt` parameter be omitted?

AlekseyTs replied:
I think calling MakeFieldAccess is unnecessary.

jcouv replied:
Makes sense. Done.

jcouv commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -806,171 +959,194 @@ class C
+                    model.GetDeclaredSymbol(x).ToTestDisplayString());
+            };
+
+            var verifier = CompileAndVerify(source, expectedOutput: @"1 2 3 4 5 6 7 Alice 2 3 4 5", additionalRefs: new[] { MscorlibRef });
```
I forgot to hook up the validator here...

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +10,65 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+                    indexInLink = fieldRemainder;
+                }
+
+                WellKnownMember wellKnownTypeMember = TupleTypeSymbol.GetTupleTypeMember(containingLinkSize, indexInLink);
```
If we keep the loop only to deal with nesting, this could be moved out of the loop because the nesting always happens through the same field.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +10,65 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+                WellKnownMember wellKnownTypeMember = TupleTypeSymbol.GetTupleTypeMember(containingLinkSize, indexInLink);
+                var linkField = (FieldSymbol)Binder.GetWellKnownTypeMember(_compilation, wellKnownTypeMember, _diagnostics, syntax: node.Syntax);
+
+                FieldSymbol nestedFieldSymbol = linkField?.AsMember(currentLinkType);
```
Conditional access is not necessary, if we check for field presence earlier.

AlekseyTs replied:
I mean during binding.

jcouv replied:
Ok

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,49 @@ internal partial class LocalRewriter
+                var ctor = (MethodSymbol)Binder.GetWellKnownTypeMember(_compilation, TupleTypeSymbol.GetTupleCtor(currentArity), _diagnostics, syntax: node.Syntax);
+                if ((object)ctor == null)
+                {
+                    return node;
```
Perhaps should mark it with HasErrors before returning.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,49 @@ internal partial class LocalRewriter
+                }
+
+                // PERF We are starting with the most nested type, which will involve multiple walking through the underlying tuple type. But long tuples are rarer.
+                NamedTypeSymbol nestedUnderlyingType = TupleTypeSymbol.GetNestedTupleType(underlyingTupleType, linkIndex);
```
As we discussed offline, this can be simplified if we push the chain of types into a builder before we enter the loop.

AlekseyTs replied:
This can be done in one pass too.

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -71,33 +72,83 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
+
+            for (int linkIndex = chainLength - 2; linkIndex >= 0; linkIndex--)
+            {
+                NamedTypeSymbol chainedTupleType = binder.GetWellKnownType(GetTupleType(8), diagnostics, syntax);
```
This can be moved out of the loop, assuming we add an 'if' to execute the assignment and the loop only when there  is a nesting.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -200,32 +254,65 @@ private static WellKnownMember GetTupleTypeMember(int arity, int position)
+            {
+                string tail = name.Substring(4);
+                int number;
+                if (int.TryParse(tail, out number))
```
As @cston mentioned, we should probably use an overload with style parameter.

jcouv replied:
I didn't find a suitable solution with `TryParse` alone, so I'm doing a round-trip.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -692,29 +779,55 @@ internal override ImmutableArray<NamedTypeSymbol> GetInterfacesToEmit()
+        internal TupleFieldSymbol WithUnderlyingType(TupleTypeSymbol newContainingTuple, NamedTypeSymbol newContainingUnderlyingType)
         {
-            get
+            FieldSymbol newUnderlyingField = _underlyingField.OriginalDefinition.AsMember(newContainingUnderlyingType);
```
It looks like _underlyingField could be null in some error cases.

jcouv replied:
Fixed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -692,29 +779,55 @@ internal override ImmutableArray<NamedTypeSymbol> GetInterfacesToEmit()
+
+            WellKnownMember wellKnownTypeMember = TupleTypeSymbol.GetTupleTypeMember(containingLinkSize, fieldRemainder);
+            var linkField = (FieldSymbol)Binder.GetWellKnownTypeMember(binder.Compilation, wellKnownTypeMember, diagnostics, syntax: syntax);
+            NamedTypeSymbol linkType = TupleTypeSymbol.GetNestedTupleType(containingUnderlyingType, fieldChainLength - 1);
```
Shouldn't probably do this if linkField is null.

jcouv replied:
Fixed.
But I will keep the general error handling change for symbol construction to a separate commit.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -692,29 +779,55 @@ internal override ImmutableArray<NamedTypeSymbol> GetInterfacesToEmit()
         private readonly TypeSymbol _type;
         private readonly TupleTypeSymbol _containingTuple;
+        private readonly int _index;
+        private readonly FieldSymbol _underlyingField;
```
Consider keeping the Opt suffix.

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -746,15 +890,21 @@ static void Main()
             var comp = CreateCompilationWithMscorlib(source);
             comp.VerifyDiagnostics(
-                // (6,50): error CS8201: Tuple member name 'Item1' is disallowed at position 2.
+                // (6,28): error CS8201: Tuple member name 'Item1' is only allowed at position 2.
```
The message is misleading, Item1 is allowed only at position **1**

jcouv replied:
Yes, I noticed that too after updating the PR. Fixed.

AlekseyTs commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -746,15 +890,21 @@ static void Main()
                 //         (int Item1, string Item1) x = (Item1: 1, Item1: "hello");
                 Diagnostic(ErrorCode.ERR_TupleReservedMemberName, "Item1").WithArguments("Item1", "2").WithLocation(6, 50),
-                // (7,40): error CS8201: Tuple member name 'Item2' is disallowed at position 1.
+                // (7,14): error CS8201: Tuple member name 'Item2' is only allowed at position 1.
```
Same issue with position.

AlekseyTs commented:
Done with review pass

jcouv commented:
@AlekseyTs @VSadov @cston Latest commits addressed the feedback we discussed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -628,110 +629,72 @@ private BoundExpression BindExpressionInternal(ExpressionSyntax node, Diagnostic
 
-            return new BoundTupleCreationExpression(node, ctor, boundArguments.ToImmutableAndFree(), type, hasErrors);
+            diagnostics.AddRange(bag);
+            return new BoundTupleCreationExpression(node, boundArguments.ToImmutableAndFree(), type, bag.HasAnyErrors());
```
Calling HasAnyErrors during initial binding is a bad idea, it resolves lazy diagnostics, which, if we do this too early, can cause cycles. Using it as a way of calculating HasErrors is also wrong because we should never mark nodes as bad because obsolete errors. Please go back to the original way of tracking HasErrors. We can talk more offline about when we should mark nodes with that flag.

jcouv replied:
Done. 
We can discuss today. I didn't understand why the local diagnostic bag approach was good for iterator rewriter, but not for local lowering.

jcouv commented:
Reverted to `hasErrors` in last commit.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +11,73 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+                if ((object)linkField == null)
+                {
+                    return node;
+                }
```
All six lines at the beginning of this loop can be moved out of the loop. Surround them and the loop with ```if``` checking that we will have to do at least one iteration and , perhaps, convert the loop to the do-while.

jcouv replied:
Done, but kept the `for` loop.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -23,6 +19,9 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol, ITupleTypeSymbol
         private readonly NamedTypeSymbol _underlyingType;
         private readonly ImmutableArray<TupleFieldSymbol> _fields;
 
+        internal const int chunk7 = 7; // Used for zero-based computation (array indices and such)
```
It doesn't look like the name follows the esteblished style. I think internal static field should start with capital letter.

jcouv replied:
Done. Thanks

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +11,73 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+                currentLinkType = TupleTypeSymbol.GetNestedTupleType(currentLinkType, 1);
+            }
+
+            {
```
A block like that looks strange, consider removing.

jcouv replied:
Done.
The main advantage was to have the parallel between code in the loop and repeated here. But with other changes this won't make sense anymore. I removed the block and renamed locals to distinguish.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +11,73 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+                int tupleRemainder;
+                int tupleChainLength = TupleTypeSymbol.NumberOfValueTuples(tupleType.Arity, out tupleRemainder);
+
+                int containingLinkSize = (tupleChainLength == fieldChainLength) ? tupleRemainder : TupleTypeSymbol.chunk8;
```
containingLinkSize is the arity of the currentLinkType, there is no need to call NumberOfValueTuples and have all this conditional logic.

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +11,73 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+
+                FieldSymbol nestedFieldSymbol = linkField.AsMember(currentLinkType);
+
+                TypeSymbol nestedFieldType = currentLinkType.TypeArguments[fieldRemainder - 1];
```
This is unnecessary, the field type is known and is node.Type.

jcouv replied:
Sure. This was an artifact of keeping the code inside the loop and outside the loop similar.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,80 @@ internal partial class LocalRewriter
+            while (true)
+            {
+                underlyingTupleTypeChain.Add(currentType);
+                if (currentType.TypeArguments.Length == TupleTypeSymbol.chunk8)
```
Arity vs. TypeArguments.Length

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,80 @@ internal partial class LocalRewriter
+                underlyingTupleTypeChain.Add(currentType);
+                if (currentType.TypeArguments.Length == TupleTypeSymbol.chunk8)
+                {
+                    currentType = (NamedTypeSymbol)currentType.TypeArguments[TupleTypeSymbol.chunk7];
```
TypeArguments is banned inside compiler, it looks like we missed this in analyzer. TypeArgumentsNoUseSiteDiagnostics should be used instead if you are not interested in use-site diagnostics, and TypeArgumentsWithDefinitionUseSiteDiagnostics if you are. The goal to make this choice explicit.

jcouv replied:
Done. Thanks
Filed issue #10265 for the analyzer problem.

jcouv replied:
I will also fix other instances of this.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,80 @@ internal partial class LocalRewriter
+            NamedTypeSymbol underlyingTupleType = ((TupleTypeSymbol)node.Type).UnderlyingTupleType;
+            ArrayBuilder<NamedTypeSymbol> underlyingTupleTypeChain = GetUnderlyingTypeChain(underlyingTupleType);
+
+            var currentCreation = default(BoundObjectCreationExpression);
```
No need to declare the local until we are ready to assign. In any case, no need to initialize it with default.

jcouv replied:
Done.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +11,73 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+                FieldSymbol nestedFieldSymbol = linkField.AsMember(currentLinkType);
+
+                TypeSymbol nestedFieldType = currentLinkType.TypeArguments[fieldRemainder - 1];
+                currentField = new BoundFieldAccess(node.Syntax, currentField, nestedFieldSymbol, node.ConstantValue, node.ResultKind, nestedFieldType);
```
return ?

jcouv replied:
Yes

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +11,73 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+            int fieldRemainder;
+            int fieldChainLength = TupleTypeSymbol.NumberOfValueTuples(tupleField.Position, out fieldRemainder);
+
+            BoundExpression currentField = rewrittenReceiver;
```
The name of this local is misleading, in fact this local is not needed, use rewrittenReceiver in its place.

jcouv replied:
Ok

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,80 @@ internal partial class LocalRewriter
+
+            var currentCreation = default(BoundObjectCreationExpression);
+
+            {
```
Consider getting rid of this scope.

jcouv replied:
Done.

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,80 @@ internal partial class LocalRewriter
+                var ctor = (MethodSymbol)Binder.GetWellKnownTypeMember(_compilation, TupleTypeSymbol.GetTupleCtor(remainder), _diagnostics, syntax: node.Syntax);
+                if ((object)ctor == null)
+                {
+                    underlyingTupleTypeChain.Free();
```
Consider calling ```underlyingTupleTypeChain.Free()``` once, in a `finally` block.

jcouv replied:
I'm confused. During the iterator rewriter work, the [recommendation](https://github.com/dotnet/roslyn/pull/9536#discussion_r56541549) was to avoid `finally` for such freeing.
Personally, I would prefer `finally` as well, as it will avoid mistakes (forgetting to free). Let me know.

cston replied:
Sorry, I should have given more detail previously. The recommendation is to avoid ```try/finally``` in cases where it is not necessary to dispose the resource (e.g.: return a builder to the pool) when an `Exception` is thrown _and_ where removing the `finally` does not introduce unnecessary complexity to the code (e.g.: multiple exit points that require disposing).

jcouv replied:
Sounds good. I'll use `finally` here.

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,80 @@ internal partial class LocalRewriter
+
+        private ArrayBuilder<NamedTypeSymbol> GetUnderlyingTypeChain(NamedTypeSymbol underlyingTupleType)
+        {
+            var underlyingTupleTypeChain = ArrayBuilder<NamedTypeSymbol>.GetInstance();
```
Should be allocated by the caller since the caller is responsible for freeing.

jcouv replied:
Ok

cston commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,80 @@ internal partial class LocalRewriter
+            return currentCreation;
+        }
+
+        private ArrayBuilder<NamedTypeSymbol> GetUnderlyingTypeChain(NamedTypeSymbol underlyingTupleType)
```
`static`?

jcouv replied:
Done

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,80 @@ internal partial class LocalRewriter
+                {
+                    underlyingTupleTypeChain.Free();
+                    return node;
+                }
```
The five lines above can be moved out of the loop. Surround them with an ```if``` checking that we have at least one iteration to perform and convert the loop to do-while.

jcouv replied:
Makes sense. Done

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,80 @@ internal partial class LocalRewriter
+        private BoundNode RewriteTupleCreationExpression(BoundTupleCreationExpression node, ImmutableArray<BoundExpression> rewrittenArguments)
+        {
+            int remainder;
+            int chainLength = TupleTypeSymbol.NumberOfValueTuples(rewrittenArguments.Length, out remainder);
```
It feels that the call to NumberOfValueTuples is unnecessary. The chainLength is amount of types in the underlyingTupleTypeChain builder after GetUnderlyingTypeChain call. remainder is the Arity of the type we are about to construct.

jcouv replied:
Done. I can indeed drive the logic from the chain in `underlyingTupleTypeChain`.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -43,21 +42,25 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol, ITupleTypeSymbol
-                                                              return new TupleFieldSymbol(field.Name, this, newUnderlyingField.Type, newUnderlyingField);
-                                                          });
+
+            _fields = originalTuple._fields.SelectAsArray(field => field?.WithUnderlyingType(this, newUnderlyingType));
```
The conditional access doesn't feel right. How can we end up with a null field here?

jcouv replied:
This is a ripple effect. The origin is in `TupleFieldSymbol.GetUnderlyingField` which may now return null if we have a problem with the well-known type for one of the ValueTuple type members.

AlekseyTs replied:
We should never have nulls in _fields array.

jcouv replied:
Ok. I'll change back and handle the error case in `GetUnderlyingField` differently.

AlekseyTs commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -396,32 +397,101 @@ internal Symbol BindNamespaceOrTypeOrAliasSymbol(ExpressionSyntax syntax, Diagno
             return tuple;
         }
 
+        private static void CollectTupleFieldMemberNames(string name, int position, int tupleSize, ref ArrayBuilder<string> elementNames)
```
Is ```position``` an index or position?

jcouv replied:
Good catch. I updated the code to indeed treat it as position.

cston commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -43,21 +42,25 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol, ITupleTypeSymbol
+        /// <summary>
+        /// Copy this tuple, but modify it to use the new underlying type.
+        /// </summary>
+        internal TupleTypeSymbol WithUnderlyingType(NamedTypeSymbol newUnderlyingType)
```
Is this method used?

jcouv replied:
Yes it is, in `AbstractTypeMap.cs` at line 56. My understanding is this is during type substitution.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -297,7 +385,7 @@ public override ImmutableArray<Symbol> GetMembers(string name)
             //PROTOTYPE: PERF do we need to have a dictionary here?
             //      tuples will be typically small 2 or 3 elements only
-            return ImmutableArray<Symbol>.CastUp(_fields).WhereAsArray(field => field.Name == name);
+            return ImmutableArray<Symbol>.CastUp(_fields).WhereAsArray(field => (object)field != null && field.Name == name);
```
The added null check is unnecessary.

jcouv replied:
`field` could be null as per comment above.

AlekseyTs replied:
_fields array should never have nulls

jcouv replied:
Reverted.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -415,7 +503,9 @@ private bool Equals(TupleTypeSymbol other, bool ignoreCustomModifiers, bool igno
-                    if (fields[i].Name != otherFields[i].Name)
+                    if ((object)fields[i] == null ||
+                        (object)otherFields[i] == null ||
+                        fields[i].Name != otherFields[i].Name)
```
The added null checks are unnecessary. Even if they were necessary, nulls on both sides should have been considered equal

jcouv replied:
The reason I added this code is to handle when one side is null. I'll stop by to discuss whether null check is necessary.

jcouv replied:
Reverted.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -545,7 +635,7 @@ public override IEnumerable<string> MemberNames
             get
             {
-                return _fields.Select(f => f.Name);
+                return _fields.Select(f => f?.Name).WhereNotNull();
```
The change is unnecessary.

jcouv replied:
Reverted.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -776,28 +779,31 @@ internal override ImmutableArray<NamedTypeSymbol> GetInterfacesToEmit()
         /// but caller is supposed to report some kind of error at declaration.
         ///
-        /// index is 0 for Item1.
+        /// index is 1 for Item1.
```
position?

cston commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -396,32 +397,101 @@ internal Symbol BindNamespaceOrTypeOrAliasSymbol(ExpressionSyntax syntax, Diagno
+            }
+        }
+
+        private static void CheckTupleMemberName(string name, int position, CSharpSyntaxNode syntax, DiagnosticBag diagnostics, PooledHashSet<string> uniqueFieldNames, ref bool hasErrors)
```
`hasErrors` is not used by the callers.

cston replied:
Consider returning `bool`, with `true` for valid tuple member name.

jcouv replied:
If it's ok with you, I'm leaving as-is because I wouldn't use the return value.

jcouv replied:
I understand the proposal now. Will do.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -805,8 +811,13 @@ internal TupleFieldSymbol(string name, TupleTypeSymbol containingTuple, TypeSymb
-            return new TupleFieldSymbol(_name, newContainingTuple, newUnderlyingField.Type, _index, newUnderlyingField);
+            if ((object)_underlyingFieldOpt == null)
+            {
+                return null;
```
This is wrong an many levels.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleFieldSymbol.cs:
```diff
@@ -0,0 +1,230 @@
+        /// A tuple without backing fields is usable for binding purposes, since we know its name and type,
+        /// but caller is supposed to report some kind of error at declaration.
+        ///
+        /// index is 1 for Item1.
```
position?

jcouv replied:
Yes, I changed from index to position.

AlekseyTs replied:
And the comment still calls it index

jcouv replied:
Fixed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleFieldSymbol.cs:
```diff
@@ -0,0 +1,230 @@
+        {
+            if ((object)_underlyingFieldOpt == null)
+            {
+                return null;
```
This is wrong on many levels.

jcouv replied:
I'll stop by to discuss null handling. Let me know when you have a minute.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleFieldSymbol.cs:
```diff
@@ -0,0 +1,230 @@
+
+        /// <summary>
+        /// Returns the index of this field within the tuple.
+        /// For instance, the index for Item1 is 1.
```
position?

jcouv replied:
I didn't understand. I changed from index (zero-based) to a position (one-based).

AlekseyTs replied:
And the comment still calls it index

jcouv replied:
Fixed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -71,33 +74,83 @@ internal static TupleTypeSymbol ConstructTupleTypeSymbol(TupleTypeSymbol origina
-            return underlyingType;
+            for (int linkIndex = chainLength - 2; linkIndex >= 0; linkIndex--)
+            {
+                NamedTypeSymbol chainedTupleType = binder.GetWellKnownType(GetTupleType(chunk8), diagnostics, syntax);
```
This can be moved out of the loop, assuming we add an 'if' to execute the assignment and the loop only when there is a nesting.

AlekseyTs replied:
There are about two suggestions of similar nature elsewhere, please look at them too.

jcouv replied:
Yes. I think I removed all instances of un-varying locals out of the loops. Thanks

AlekseyTs commented:
Done with review pass

jcouv commented:
@AlekseyTs @VSadov @cston Latest commit addresses the feedback we discussed. Thanks again for the feedback.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -10,6 +10,7 @@
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.Text;
 using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.Collections;
```
Sort usings

jcouv replied:
Will do. Makes me wonder if we could fix the "add using" fixer to insert differently in the first place.

jaredpar replied:
It seems to be 50/50 for me if it sorts or not.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -9,6 +9,7 @@
 using Microsoft.CodeAnalysis.CSharp.Syntax;
 using Microsoft.CodeAnalysis.RuntimeMembers;
 using Roslyn.Utilities;
+using Microsoft.CodeAnalysis.Collections;
```
Sort usings

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -396,32 +397,101 @@ internal Symbol BindNamespaceOrTypeOrAliasSymbol(ExpressionSyntax syntax, Diagno
+            ImmutableArray<TypeSymbol> typesArray = types.ToImmutableAndFree();
+            if (typesArray.Length < 2)
+            {
+                // PROTOTYPE(tuples)
```
Shouldn't we also be adding a diagnostic here?  Or is that what this PROTOTYPE instance is tracking?

jcouv replied:
It should not be possible to reach this as there is no language syntax for 1-tuples or 0-tuples. But I left the PROTOTYPE to sort out the proper error handling.

jaredpar commented on src/Compilers/CSharp/Portable/Binder/Binder_Symbols.cs:
```diff
@@ -396,32 +397,101 @@ internal Symbol BindNamespaceOrTypeOrAliasSymbol(ExpressionSyntax syntax, Diagno
+                typesArray,
+                elementNames == null ?
+                    default(ImmutableArray<string>) :
+                    elementNames.ToImmutableAndFree(),
```
It seems like it's possible here for `typesArray` and `elementNames` to have different lengths.  The length of `typesArray` will always be `numElements`.  The `elementNames` can have a different count, as noted on line 427.  In that case we raise an error but still create a `TupleTypeSymbol` where the arrays have different lengths.  

Won't that make the `TupleTypeSymbol` more difficult to program against?  It would mean developers have to always account for the case where the type array and name array have different lengths.  A solution could be to clear `elementNames` in the case we raise `ERR_TupleExplicitNamesOnAllMembersOrNone`.

jcouv replied:
When we bind tuples (declaration and creation), we currently place a name for every element whether or not the user put one. So currently they always match. 
I will add an assert since we expect them to be of equal lengths at this point.

That said, we discussed maybe not populating Item1 through 7 if the user didn't provide names (and let the underlying type handling them). 
I'm not sure I favor that, but if we did that, then we could end up with fewer element names than types.

jaredpar commented on src/Compilers/CSharp/Portable/Symbols/TupleFieldSymbol.cs:
```diff
@@ -0,0 +1,225 @@
+        internal TupleFieldSymbol WithUnderlyingType(TupleTypeSymbol newContainingTuple, NamedTypeSymbol newContainingUnderlyingType)
+        {
+            FieldSymbol newUnderlyingFieldOpt = _underlyingFieldOpt?.OriginalDefinition.AsMember(newContainingUnderlyingType);
+            return new TupleFieldSymbol(_name, newContainingTuple, newUnderlyingFieldOpt.Type, _position, newUnderlyingFieldOpt);
```
Appears that `newUnderlyingFieldOpt` can be null here hence `newUnderlyingFieldOpt.Type` is subject to NRE.

jcouv replied:
What is NRE?

jcouv replied:
Nevermind, got it, null ref exception.

jcouv replied:
I'm looking into this problem.

jcouv replied:
This is now fixed in latest commit. It also includes a test to cover this case for long tuples.
Thanks

VSadov commented on src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenTupleTest.cs:
```diff
@@ -806,171 +996,193 @@ class C
 
         [Fact]
-        public void Tuple2To7Members()
+        public void LongTupleCreationWithNames()
```
I think we also need a test confirming the order of argument evaluation. It should be left to right. The easiest way is to execute something like:
```C#
 var t = (a1: PrintAndReturn(1), a2: PrintAndReturn(2),... , a18: PrintAndReturn(18)); 
```

jcouv replied:
Added such test in latest commit.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_Field.cs:
```diff
@@ -12,11 +11,65 @@ public override BoundNode VisitFieldAccess(BoundFieldAccess node)
+                {
+                    FieldSymbol nestedFieldSymbol = tupleRestField.AsMember(currentLinkType);
+
+                    currentLinkType = (NamedTypeSymbol)currentLinkType.TypeArgumentsNoUseSiteDiagnostics[TupleTypeSymbol.Chunk7];
```
The purpose of having named constants is to use meaningful names. I don't see this for Chunk7 and Chunk8. Not only names are meaningless to me, if feels like they are used in some places just because they happened to have a certain value rather than because they have meaning related to operation being performed. Like on this line for example. Perhaps we should have more than one constant with the same value, but with unique meaning. For example, a constant used in GetTupleTypeMember call above, could have name ExpanderFieldPosition. A constant on this line could be named ExpanderFieldIndex.

jcouv replied:
That's fair. 
Initially, the purpose was to just capture the magic number 8, so that we could expand our library of known tuples without chasing them. Then I realized I don't want to do `8 -1` in lots of places. 
Basically, `8` is the max size of ValueTuple and `7` is the same thing as zero-based. I'm open to better names, but I don't think we need to introduce more flavors.

AlekseyTs commented on src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleCreationExpression.cs:
```diff
@@ -13,8 +9,68 @@ internal partial class LocalRewriter
+                // make a creation expression for the smallest type
+                NamedTypeSymbol smallestType = underlyingTupleTypeChain.Pop();
+                ImmutableArray<BoundExpression> smallestCtorArguments = ImmutableArray.Create(rewrittenArguments,
+                                                                                              underlyingTupleTypeChain.Count * TupleTypeSymbol.Chunk7,
```
Cunk7 on this line, for example, could be called MaxNumberOfFieldsInOneChunk.

jcouv replied:
Removed `Chunk7`. Renamed `Chunk8` as we discussed.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleFieldSymbol.cs:
```diff
@@ -0,0 +1,225 @@
+        internal TupleFieldSymbol WithType(TupleTypeSymbol newContainingTuple, NamedTypeSymbol newlinkType, TypeSymbol newFieldType)
+        {
+            FieldSymbol newUnderlyingFieldOpt = _underlyingFieldOpt?.OriginalDefinition.AsMember(newlinkType);
+            return new TupleFieldSymbol(_name, newContainingTuple, newFieldType, _position, newUnderlyingFieldOpt);
```
Consider adding an assert that newUnderlyingFieldOpt.Type == newFieldType when newUnderlyingFieldOpt exists.

jcouv replied:
Agreed.

jcouv replied:
Done. 
I wasn't 100% if there are different kinds of equality I should worry about. I just used `==` operator as you suggested.

AlekseyTs commented on src/Compilers/CSharp/Portable/Symbols/TupleTypeSymbol.cs:
```diff
@@ -43,21 +42,66 @@ internal sealed class TupleTypeSymbol : NamedTypeSymbol, ITupleTypeSymbol
-                                                          });
+
+            ArrayBuilder<NamedTypeSymbol> underlyingTupleTypeChain = ArrayBuilder<NamedTypeSymbol>.GetInstance();
+            GetUnderlyingTypeChain(newUnderlyingType, underlyingTupleTypeChain);
```
We could probably do without collecting all the types into the builder. The types are needed in the same order, as they are chained in type arguments, so we could just dig when appropriate as we iterate through the fields.

jcouv replied:
Done.

AlekseyTs commented:
Done with review pass.

jcouv commented:
Pushed another commit which addresses the naming of `Chunk8` and tightens the loop for substituting underlying type.

AlekseyTs commented:
LGTM

VSadov commented:
LGTM
